feat: HTTP API specification for opportunities, suggestions

iuliag opened the pull request:
Fixes https://jira.corp.adobe.com/browse/SITES-26859

HTTP API specification the opportunities and suggestions following the new data model introduced in https://adobe.enterprise.slack.com/docs/T025LQ61T/F07U2CQ4QH3, https://jira.corp.adobe.com/browse/SITES-26550

Includes for starters also:
- examples for broken backlinks opportunity and suggestion
- schemas for:
  - broken backlinks guidance
  - broken backlinks redirect data
  - broken backlinks kpisDelta
  - broken backlinks opportunity data

That can be used with `oneOf` for those properties that need customization per opportunity or suggestion type.

![opportunity](https://github.com/user-attachments/assets/4b35d804-23a7-4fbd-a897-982afaec082e)
![suggestion](https://github.com/user-attachments/assets/d1ec0638-7dd1-49f2-b17e-c32b3ac2b272)

Fixes will follow in another PR.
Implementation will following via https://jira.corp.adobe.com/browse/SITES-26860.

Please ensure your pull request adheres to the following guidelines:
- [x] make sure to link the related issues in this description. Or if there's no issue created, make sure you 
  describe here the problem you're solving.
- [x] when merging / squashing, make sure the fixed issue references are visible in the commits, for easy compilation of release notes

If the PR is changing the API specification:
- [x] make sure you add a "Not implemented yet" note the endpoint description, if the implementation is not ready 
  yet. Ideally, return a 501 status code with a message explaining the feature is not implemented yet.
- [x] make sure you add at least one example of the request and response.

If the PR is changing the API implementation or an entity exposed through the API:
- [ ] make sure you update the API specification and the examples to reflect the changes.

If the PR is introducing a new audit type:
- [ ] make sure you update the API specification with the type, schema of the audit result and an example

## Related Issues


Thanks for contributing!

solaris007 commented on docs/openapi/schemas.yaml:
```diff
@@ -1598,3 +1598,313 @@ ApiKeyResponse:
+    origin:
+      description: Origin of opportunity
+      type: string
+      enum:
```
we also have AUTOMATION here for opportunities created from audits, but not AI

solaris007 commented on docs/openapi/schemas.yaml:
```diff
@@ -1598,3 +1598,313 @@ ApiKeyResponse:
+    - title
+    - status
+    - createdAt
+    - createdBy
```
createdBy and updatedBy will not be initially available and be derived automatically from auth context in the future

iuliag replied:
This is the entity for the responses. I let the property still be required, but included this mention in the description of it. Would that be fine or do you want to have it required only when we'll be able to derive it?

solaris007 replied:
yes, for clarity purposes and to prevent wrong expectations, i'd leave it out for now.

iuliag replied:
Okay, I removed them.

solaris007 reviewed (commented):
security should be ims_key everywhere - api keys deprecated

ekremney commented:
@iuliag it is very hard to visualize the complex YAML schemas with lots of inter-references. Would you mind adding screenshots from OpenAPI UI or sharing example JSONs for opportunity and suggestion schemas for people to better grasp what's being proposed? 

here is an example: https://github.com/adobe/spacecat-api-service/pull/574

rpapani commented on docs/openapi/api.yaml:
```diff
@@ -105,6 +105,20 @@ paths:
+    $ref: './site-opportunities.yaml#/site-opportunity-suggestions-by-status'
+  /sites/{siteId}/opportunities/{opportunityId}/suggestions/{suggestionId}:
+    $ref: './site-opportunities.yaml#/site-opportunity-suggestion'
+  /sites/{siteId}/opportunities/{opportunityId}/suggestions/status:
```
what is the difference between
`/sites/{siteId}/opportunities/{opportunityId}/suggestions/by-status/{status}:`
vs
`/sites/{siteId}/opportunities/{opportunityId}/suggestions/status:`

iuliag replied:
`GET /sites/{siteId}/opportunities/{opportunityId}/suggestions/by-status/{status}` lists suggestions by status
`PATCH /sites/{siteId}/opportunities/{opportunityId}/suggestions/status` updates the status of one or more suggestions

rpapani replied:
wouldn't it make sense to use the same path for both GET and PUT/PATCH?

iuliag replied:
Then I'd leave the `GET /sites/{siteId}/opportunities/{opportunityId}/suggestions/by-status/{status}` as is, because it aligns with other get list of filtered entities APIs.
And turn `PATCH /sites/{siteId}/opportunities/{opportunityId}/suggestions/status` into `PATCH /sites/{siteId}/opportunities/{opportunityId}/suggestions` with the only property allowed to update for now being status.
@solaris007 wdyt?

solaris007 replied:
that is confusing - previously the URL made clear that we're updating status. like this we're obfuscating? i'd leave it as it is.

iuliag commented on docs/openapi/schemas.yaml:
```diff
@@ -1598,3 +1598,313 @@ ApiKeyResponse:
+      description: Origin of opportunity
+      type: string
+      enum:
+        - 'ESS_OPS'
```
Yes, I updated the description with the meaning of each.

rpapani commented on docs/openapi/schemas.yaml:
```diff
@@ -1598,3 +1598,314 @@ ApiKeyResponse:
+      type: string
+      format: url
+      example: 'https://example.com/runbook/broken-backlinks'
+    type:
```
can we've `rank` property for opportunity entity as well? that'll make it easy for UI to show only the top opportunities for each opportunity type

iuliag replied:
^ @solaris007 
I see in the mocks there is a "sort by" in the opportunities listing. One of the options is "Highest impact", now if we have rank: 800000 (organic traffic) for one opportunity with rank: 14000 (click rate) for another opportunity, not sure which one is the highest impact and how this could be shown in the UI.

solaris007 replied:
@rpapani @iuliag as Iulia pointed out, there is no intention to introduce a rank for opportunities, due to type-specific dimensions. many opportunities will be singletons anyways and other than that i'd sort by status and updated at or number of open issues (suggestions)...

rpapani replied:
@solaris007 @iuliag 
we'll introduce something for now specific to opportunity type. For example, for `high-organic-low-ctr/high-inorganic-high-bourncerate`, we'd like to introduce, potentialClicks, which is computed as follows

```
potentialClicks = (siteWideCTR - pageCTR) * 100 * pageviews
```
Since these opportunities are trying to increase the engagement, this potentialClicks correctly measures what can be achieved and we'll give 10 such opportunities and UI can sort accordingly.

cc: @ekremney @shrotia

solaris007 replied:
sure, we can consider adding a rank. let us first merge the current state of the data model

iuliag commented:
> @iuliag it is very hard to visualize mentally the complex YAML schemas with lots of inter-references. Would you mind adding screenshots from OpenAPI UI or sharing example JSONs for opportunity and suggestion schemas for people to better grasp what's being proposed?
> 
> here is an example: #574

Screenshots uploaded.
The examples are available in: https://github.com/adobe/spacecat-api-service/pull/556/files#diff-57c406bcab7ca69c8aa31cd4c2770ff2476bdb6e2219a157b0f4a8b40edea729R1 and don't have inter-references.

chicharr commented on docs/openapi/schemas.yaml:
```diff
@@ -1598,3 +1598,381 @@ ApiKeyResponse:
+    siteId:
+      description: UUID of the site this opportunity is for
+      $ref: '#/Id'
+    auditId:
```
Don't know why, but `auditId` is missing from the docs.
Is it a request body parameter or a request path parameter?
![Screenshot 2024-11-15 at 16 51 06](https://github.com/user-attachments/assets/c958280e-4fab-406e-9f67-8c6b5689ad30)

iuliag replied:
It's a request body parameter.
I will check on Monday where it got lost.

iuliag replied:
Found. The referenced `Id` schema had `readOnly: true`, so the property wasn't rendered in request body. We needed to explicitly set `readOnly: false` for the property in the request body schema.

chicharr commented on docs/openapi/schemas.yaml:
```diff
@@ -1598,3 +1598,381 @@ ApiKeyResponse:
+  items:
+    type: object
+    properties:
+      id:
```
This id is also lost :) 
<img width="1173" alt="Screenshot 2024-11-15 at 19 10 03" src="https://github.com/user-attachments/assets/1e5c776b-2e25-43a6-9b90-173f14ce75fc">

iuliag replied:
Found. The referenced Id schema had readOnly: true, so the property wasn't rendered in request body. We needed to explicitly set readOnly: false for the property in the request body schema.

rpapani commented on docs/openapi/schemas.yaml:
```diff
@@ -1598,3 +1598,381 @@ ApiKeyResponse:
+  description: Type of suggestion
+  type: string
+  enum:
+    - 'redirect'
```
@iuliag 
can we add the following as well for `SuggestionType` or similar
```
experiment - Run the experiment with the given suggestions
content-change - change the content to the suggested
```

iuliag replied:
Added.

iuliag commented:
Incorporated schema and example for
- bulk create suggestions:
```
{
  "suggestions": [
    {
      "id": "2563e4567-f543-12d3-a456-426614174000",
      "opportunityId": "123e4567-e89b-12d3-a456-426614174000",
      "type": "REDIRECT_UPDATE",
      "status": "NEW",
      "rank": 571000,
      "data": {
        "referringPageTitle": "Black Friday Sale at Example Inc.",
        "referringURL": "https://www.dailynews.com/2024/08/13/black-friday-sale-at-example-inc/",
        "brokenTargetURL": "https://www.example.com/CustomerServiceView.aspx",
        "proposedTargetURLs": [
          "https://www.example.com/customer-service/contact",
          "https://www.example.com/customer-service"
        ],
        "referringPageTraffic": 571000
      },
      "createdAt": "2024-01-22T12:00:00Z",
      "updatedAt": "2024-07-26T12:00:00Z"
    },
    {
      "suggestionIndex": 1,
      "statusCode": "400",
      "message": "Validation failed in suggestion: type is required"
    }
  ],
  "metadata": {
    "total": 2,
    "success": 1,
    "failure": 1
  }
}
```

- bulk update suggestions status:
```
{
  "suggestions": [
    {
      "id": "213e4567-f543-12d3-a456-426614174000",
      "opportunityId": "123e4567-e89b-12d3-a456-426614174000",
      "type": "REDIRECT_UPDATE",
      "rank": 571000,
      "data": {
        "referringPageTitle": "Black Friday Sale at Example Inc.",
        "referringURL": "https://www.dailynews.com/2024/08/13/black-friday-sale-at-example-inc/",
        "brokenTargetURL": "https://www.example.com/CustomerServiceView.aspx",
        "proposedTargetURLs": [
          "https://www.example.com/customer-service/contact",
          "https://www.example.com/customer-service"
        ],
        "manualOverrideURL": "https://www.example.com/customer-service/faq",
        "referringPageTraffic": 571000
      },
      "status": "APPROVED",
      "createdAt": "2024-01-22T12:00:00Z",
      "updatedAt": "2024-07-26T12:00:00Z"
    },
    {
      "id": "945e4567-f543-12d3-a456-426614174010",
      "statusCode": "400",
      "message": "Validation failed in suggestion: status must be one of NEW,APPROVED,SKIPPED,FIXED,ERROR"
    }
  ],
  "metadata": {
    "total": 2,
    "success": 1,
    "failure": 1
  }
}
```

solaris007 commented:
:tada: This PR is included in version 1.79.0 :tada:

The release is available on:
- `v1.79.0`
- [GitHub release](https://github.com/adobe/spacecat-api-service/releases/tag/v1.79.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
